### PR TITLE
Persistent storage to use json

### PR DESCRIPTION
**Pull request by daniellevass**
I've updated the Persistent Storage Queue to use `ObjectQueue` instead of `QueueFile` which means we also need a converter as described here - https://github.com/square/tape#converter

We are using the Moshi implementation as that's also designed by Square and so work better together. 

This means we've added two new libraries
* Moshi - https://github.com/square/moshi
* Moshi Adapters - https://github.com/square/moshi/tree/master/adapters

I've also increased our test coverage to test each type of our Server Sync Jobs and updated the ones that originally wrote Strings to write a Server Sync Job type instead. 

I've extracted the Server Sync Job types into their own file, and put the converter there so in future if we add any more Jobs, we only need to update them all in one place :-)

Finally, I've also converted our java `syncronized()` functions into kotlin `@synchronized` annotations

### CODE REVIEW

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/PersistentJobQueue.kt [hunk 1]
```diff
@@ -12,67 +12,26 @@ interface PersistentJobQueue<T: Serializable> {
...
-     try {
-      @Suppress("unchecked_cast")
-      return objectInputStream.readObject() as T
-    } catch (e: InvalidClassException) {
-       log.w("Failed to read data from tape. Continuing without this data.")
-    }
-
-    return null
+  @Synchronized override fun peek(): T? {
+    return queue.peek()
```
Comment (luismfonseca) [inline pushnotifications/src/main/java/com/pusher/pushnotifications/internal/PersistentJobQueue.kt]: What will happen if the data is corrupted? It seems like we no longer handle this here (at least explicitly).

File: pushnotifications/src/main/java/com/pusher/pushnotifications/internal/PersistentJobQueue.kt [hunk 2]
```diff
@@ -12,67 +12,26 @@ interface PersistentJobQueue<T: Serializable> {
...
-          objectInputStream.readObject() as T
-        } catch (e: InvalidClassException) {
-          log.w("Failed to read data from tape. Continuing without this data.")
-          null
-        }
-
-      }.filterNotNull().toList() // forcing the list to be computed here in its entirety
-    }
+  @Synchronized override fun asIterable(): Iterable<T> {
+    return queue.asIterable()
```
Comment (luismfonseca) [inline pushnotifications/src/main/java/com/pusher/pushnotifications/internal/PersistentJobQueue.kt]: Same as previous comment.

File: pushnotifications/src/test/java/com/pusher/pushnotifications/internal/PersistentJobQueueTest.kt [hunk 3]
```diff
@@ -17,52 +20,194 @@ class PersistentJobQueueTest {
...
+  fun `iterables returns correct Refresh Token Job`() {
+    val queue: PersistentJobQueue<ServerSyncJob> = TapeJobQueue(tempFile, converter)
+    queue.push(RefreshTokenJob("new_token"))
+
+    val returnedList = queue.asIterable().toList()
+
+    assertNotNull(returnedList)
+    assertEquals(1, returnedList.size)
+
+    assertEquals(RefreshTokenJob::class.java, returnedList[0].javaClass)
```
Comment (luismfonseca) [inline pushnotifications/src/test/java/com/pusher/pushnotifications/internal/PersistentJobQueueTest.kt]: Does this work: `assertTrue(returnedList[0] is RefreshTokenJob)`?

File: pushnotifications/src/test/java/com/pusher/pushnotifications/internal/PersistentJobQueueTest.kt [hunk 4]
```diff
@@ -17,52 +20,194 @@ class PersistentJobQueueTest {
...
+
+    val queue: PersistentJobQueue<ServerSyncJob> = TapeJobQueue(tempFile, converter)
+
+    queue.push(StartJob("fcm_token", arrayListOf("one", "two")))
+    queue.push(RefreshTokenJob("new_token"))
+    queue.push(SubscribeJob("interest_subscribe"))
+    queue.push(UnsubscribeJob("interest_unsubscribe"))
+    queue.push(SetSubscriptionsJob(setOf("subscribe_one", "subscribe_two")))
+    queue.push(ApplicationStartJob(
+            DeviceMetadata("sdk_version", "android_version")))
```
Comment (luismfonseca) [inline pushnotifications/src/test/java/com/pusher/pushnotifications/internal/PersistentJobQueueTest.kt]: I personally don't like these values as they are not realistic examples of what they will hold when used by customers. I get that using vegetables is also not 100% realistic, but at least it's using them more closely as customers would think about them. This is more in line with the principle of "the closer test-data matches production use, the better".

In this case, the sdk version would never hold a value like `"sdk_version"` which just makes it read weird to me. Anyway, this is all a bit subjective, so treat it as an optional nitpick. :)

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+- Update persistent storage layer to use JSON
```
[the remark this answers is not in the document]
Comment (daniellevass) [reply CHANGELOG.md]: Ah I follow better now! Thank you for sharing the spec!

**Comment by daniellevass** [comment]
Just a note about the last commit - i had to comment out the test where we try to de-serialize an object that doesn't exist anymore. This is because we're now logging a warning, and because we're not mocking that logger or have a test logger it's failing at that part. I've commented it out and put a //todo to uncomment after we've refactored the logger. Does that make sense?

**Pull request merged**